experimental/FTheoryTools/src/FamilyOfG4Fluxes/attributes.jl

```diff
@@ -94,3 +94,148 @@ true
+    S = cox_ring(ambient_space(m))
+    #return S
+    gS = gens(cox_ring(ambient_space(m)))
+    linear_relations = matrix(QQ, matrix(ZZ, rays(ambient_space(m))))
+    scalings = [c.coeff for c in S.d]
+    mnf = Oscar._minimal_nonfaces(ambient_space(m))
+    sr_ideal_pos = [Vector{Int}(Polymake.row(mnf, i)) for i in 1:Polymake.nrows(mnf)]
+    data = (
+      S = S,
+      gS = gS,
+      linear_relations = linear_relations,
+      scalings = scalings,
+      sr_ideal_pos = sr_ideal_pos
+    )
+  #end
+
+  # Find the number of integral and rational parameters
+  numb_int_parameters = ncols(matrix_integral(fgs))
+  numb_rat_parameters = ncols(matrix_rational(fgs))
+
+  # Create a polynomial ring with parameters ai for the integral_parameters and ri for the rational parameters
+  gens_strings = vcat([["a$(k)" for k in 1:numb_int_parameters], ["r$(k)" for k in 1:numb_rat_parameters]]...)
+  amb_ring, gens = polynomial_ring(QQ, gens_strings)
+
+  # Extract ambient space basis of G4-flux candidates used to express flux family in
+  basis = ambient_space_models_of_g4_fluxes(m, check = check)
+  basis_indices = get_attribute(m, :ambient_space_models_of_g4_fluxes_indices)::Vector{Tuple{Int64, Int64}}
+
+  # Compute the cohomology class corresponding to the hypersurface equation
+  cy = polynomial(cohomology_class(toric_divisor_class(ambient_space(m), degree(hypersurface_equation(m)))))
+
+  # Compute the tadpole constraint by iterating over all generators of the flux family
+  tadpole_constraint_polynomial = zero(amb_ring)
+  for k1 in 1:ngens(amb_ring)
+    for k2 in 1:ngens(amb_ring)
+
+      # Extract generator k1
+      if numb_int_parameters >= k1
+        gen1 = matrix_integral(fgs)[:, k1]
+      else
+        gen1 = matrix_rational(fgs)[:, k1 - numb_int_parameters]
+      end
+
+      # Extract generator k2
+      if numb_int_parameters >= k2
+        gen2 = matrix_integral(fgs)[:, k2]
+      else
+        gen2 = matrix_rational(fgs)[:, k2 - numb_int_parameters]
+      end
+
+      # Compute the intersection number of generator k1 and generator k2
+      inter_number = ZZ(0)
+      for l1 in 1:length(basis)
+        for l2 in 1:length(basis)
+
+          gen1[l1] * gen2[l2] == 0 && continue
+
+          my_tuple = Tuple(sort([basis_indices[l1]..., basis_indices[l2]...]))
+
+          if arxiv_doi(m) == "10.48550/arXiv.1511.03209"
+            change = sophisticated_intersection_product(ambient_space(m), my_tuple, hypersurface_equation(m), inter_dict, s_inter_dict, data)
+          else
+            change = get!(inter_dict, my_tuple) do
+              return QQ(integrate(cohomology_class(ambient_space(m), polynomial(basis[l1]) * polynomial(basis[l2]) * cy); check = check))
+            end        
+          end
+
+          inter_number += gen1[l1] * gen2[l2] * change
+          
+        end
+      end
+
+      # Update the D3-tadpole constraint polynomial
+      tadpole_constraint_polynomial += inter_number * gens[k1] * gens[k2]
```

Comment:
Dunno how many terms these `tadpole_constraint_polynomial` have in practice. If there are just few, this is fine. If you have many (thousands) might be worth using a `MPolyBuildCtx` to construct it?

Reply:
Yes that is definitely in order. Just wanted a first working version first. Have just updated the code accordingly.